consensus/snapshot/snapshots_service.c

```diff
@@ -0,0 +1,420 @@
+  if (target_milestone_tx_pack.num_loaded == 0) {
+    return RC_SNAPSHOT_MISSING_MILESTONE_TRANSACTION;
+  }
+
+  while (index > initial_snapshot->metadata.index && index > 1) {
+    ERR_BIND_RETURN(iota_snapshots_service_find_solid_entry_points_and_update(
+                        snapshots_service, target_milestone->index, transaction_timestamp(target_milestone_tx_p),
+                        prev_milestone.hash, prev_milestone.index, tangle, solid_entry_points),
+                    ret);
+    if ((hash_to_uint64_t_map_size(*solid_entry_points) - solid_entry_points_count) == 0) {
+      log_info(logger_id,
+               "No solid entry points found at for target milestone %" PRIu64 " under milestone %" PRIu64 "\n",
+               target_milestone->index, prev_milestone.index);
+      break;
+    }
+    ERR_BIND_RETURN(hash_to_uint64_t_map_add(solid_entry_points, prev_milestone.hash, prev_milestone.index), ret);
+    solid_entry_points_count = hash_to_uint64_t_map_size(*solid_entry_points);
+    hash_pack_reset(&prev_milestone_pack);
+    ERR_BIND_RETURN(iota_tangle_milestone_load_by_index(tangle, index - 1, &prev_milestone_pack), ret);
+    if (prev_milestone_pack.num_loaded == 0) {
+      return RC_SNAPSHOT_MISSING_MILESTONE_TRANSACTION;
+    } else {
+      index = prev_milestone.index;
+    }
+  }
+  return RC_OK;
+}
+
+retcode_t iota_snapshots_service_update_solid_entry_points(snapshots_service_t *const snapshots_service,
+                                                           snapshot_t *const snapshot,
+                                                           iota_milestone_t const *const target_milestone,
+                                                           tangle_t const *const tangle) {
+  retcode_t ret;
+  hash_to_uint64_t_map_t solid_entry_points = NULL;
+
+  iota_snapshot_read_lock(snapshot);
+  hash_to_uint64_t_map_copy(&snapshot->metadata.solid_entry_points, &solid_entry_points);
+  iota_snapshot_unlock(snapshot);
+
+  hash_to_uint64_t_map_remove(&solid_entry_points, snapshots_service->conf->genesis_hash);
+
+  ERR_BIND_GOTO(
+      hash_to_uint64_t_map_add(&solid_entry_points, snapshots_service->conf->genesis_hash, target_milestone->index),
+      ret, cleanup);
+
+  ERR_BIND_GOTO(iota_snapshots_service_update_old_solid_entry_points(
+                    snapshots_service, &snapshots_service->snapshots_provider->inital_snapshot, target_milestone,
+                    tangle, &solid_entry_points),
+                ret, cleanup);
+  ERR_BIND_GOTO(iota_snapshots_service_collect_new_solid_entry_points(
+                    snapshots_service, &snapshots_service->snapshots_provider->inital_snapshot, target_milestone,
+                    tangle, &solid_entry_points),
+                ret, cleanup);
+
+cleanup:
+  iota_snapshot_write_lock(snapshot);
+  hash_to_uint64_t_map_free(&snapshot->metadata.solid_entry_points);
```

Comment:
Put it before the lock ?

Reply:
sorry, lock is not needed here at all, _snapshot_ is a local variable of a calling function